File: BasicLevel_C++/1034. 有理数四则运算(20).cpp
#include <iostream>
#include <cmath>
using namespace std;
long long a, b, c, d;
long long gcd(long long t1, long long t2) {
    return t2 == 0 ? t1 : gcd(t2, t1 % t2);
}
void func(long long m, long long n) {
    int flag1 = 0, flag2 = 0, flag = 0;
    if (n == 0) {
        printf("Inf");
        return ;
    }
    if (m == 0) {
        printf("0");
        return ;
    }
    if (m < 0) flag1 = 1;
    if (n < 0) flag2 = 1;
    m = abs(m), n = abs(n);
    if (flag1 == 1 && flag2 == 1) flag = 0;
    else if (flag1 == 1 || flag2 == 1) flag = 1;
    if (m == n) {
        if (flag == 1) printf("(-1)");
        else printf("1");
        return;
    }       
    long long x = m % n, y = m / n;
    if (x == 0) {
        if (flag == 0) printf("%d", y);
        else printf("(-%d)", y);
        return ;
    } else {
        long long t1 = m - y * n, t2 = n, t = gcd(t1, t2);
        t1 = t1 / t, t2 = t2 / t;
        if (flag == 1) {
            printf("(-");
            if (y != 0) printf("%lld %lld/%lld)", y, t1, t2);
            else printf("%d/%d)", t1, t2);
        } else {
            if (y != 0) printf("%lld %lld/%lld", y, t1, t2);
            else printf("%lld/%lld", t1, t2);
        }
    }
}
void print() {
    func(a, b); printf(" + "); func(c, d); printf(" = "); func(a * d + b * c, b * d); printf("\n");
    func(a, b); printf(" - "); func(c, d); printf(" = "); func(a * d - b * c, b * d); printf("\n");
    func(a, b); printf(" * "); func(c, d); printf(" = "); func(a * c, b * d); printf("\n");
    func(a, b); printf(" / "); func(c, d); printf(" = "); func(a * d, b * c); printf("\n");
}
int main() {
    scanf("%lld/%lld %lld/%lld", &a, &b, &c, &d);
    print();
    return 0;
}

File: BasicLevel_Golang/1034. 有理数四则运算(20).go
package main

import (
	"fmt"
	"math"
	"strconv"
	"strings"
)

//Gcd 最大公约数：辗转相除法
func Gcd(x, y int64) int64 {
	x = int64(math.Abs(float64(x)))
	y = int64(math.Abs(float64(y)))

	var tmp int64
	for {
		tmp = (x % y)
		if tmp > 0 {
			x = y
			y = tmp
		} else {
			return y
		}
	}
}

//Lcm 最小公倍数
func Lcm(x, y int64) int64 {
	return (x * y) / Gcd(x, y)
}

//Add 分数加法
func Add(a1, b1, a2, b2 int64) (c1, c2 int64) {
	lcm := Lcm(b1, b2)
	c1 = a1*(lcm/b1) + a2*(lcm/b2)
	c2 = lcm
	return c1, c2
}

//Sub 分数减法
func Sub(a1, b1, a2, b2 int64) (c1, c2 int64) {
	lcm := Lcm(b1, b2)
	c1 = a1*(lcm/b1) - a2*(lcm/b2)
	c2 = lcm
	return c1, c2
}

//Mul 分数乘法
func Mul(a1, b1, a2, b2 int64) (c1, c2 int64) {
	gcd := Gcd(a1*a2, b1*b2)
	c1 = a1 * a2 / gcd
	c2 = b1 * b2 / gcd
	return c1, c2
}

//Div 分数除法
func Div(a1, b1, a2, b2 int64) (c1, c2 int64) {
	c1 = a1 * b2
	c2 = a2 * b1
	if (c1 < 0 && c2 < 0) || (c1 > 0 && c2 < 0) {
		c1, c2 = c1*(-1), c2*(-1)
	}
	return c1, c2
}

//Display 精简显示
func Display(c1, c2 int64) string {
	var flag int
	var str []string
	if c2 == 0 {
		return "Inf"
	}
	if c1 == 0 {
		return "0"
	}
	gcd := Gcd(c1, c2)
	if c1 < 0 {
		str = append(str, "(", "-")
		flag = 1
		c1 *= (-1)
	}
	k := c1 / c2
	a, b := c1/gcd, c2/gcd
	if k != 0 {
		str = append(str, strconv.Itoa(int(k)))
	}
	if k != 0 && b != 1 {
		str = append(str, " ")
	}
	if b != 1 {
		str = append(str, strconv.Itoa(int(a%b)), "/", strconv.Itoa(int(b)))
	}
	if flag == 1 {
		str = append(str, ")")
	}
	result := strings.Join(str, "")
	return result
}

func main() {
	var m, n string
	fmt.Scanln(&m, &n)

	str1 := strings.Split(m, "/")
	a1, _ := strconv.Atoi(str1[0])
	b1, _ := strconv.Atoi(str1[1])
	str2 := strings.Split(n, "/")
	a2, _ := strconv.Atoi(str2[0])
	b2, _ := strconv.Atoi(str2[1])
	c1 := Display(int64(a1), int64(b1))
	c2 := Display(int64(a2), int64(b2))
	s1, s2 := Add(int64(a1), int64(b1), int64(a2), int64(b2))
	fmt.Println(c1, "+", c2, "=", Display(s1, s2))
	d1, d2 := Sub(int64(a1), int64(b1), int64(a2), int64(b2))
	fmt.Println(c1, "-", c2, "=", Display(d1, d2))
	p1, p2 := Mul(int64(a1), int64(b1), int64(a2), int64(b2))
	fmt.Println(c1, "*", c2, "=", Display(p1, p2))
	q1, q2 := Div(int64(a1), int64(b1), int64(a2), int64(b2))
	fmt.Println(c1, "/", c2, "=", Display(q1, q2))
}
